src/api_call.rs

```diff
@@ -0,0 +1,143 @@
+use crate::service::Service;
+use crate::application::Application;
+use crate::user::User;
+use crate::usage::Usage;
+use crate::errors::*;
+
+use crate::ToParams;
+
+use std::collections::HashMap;
+
+#[derive(Copy, Clone, Debug)]
+pub enum Kind {
+    Authorize,
+    AuthRep,
+    Report
+}
+
+type Extensions = HashMap<String, String>;
+
+#[derive(Copy, Clone, Debug)]
+pub struct ApiCall<'service, 'app, 'user, 'usage, 'extensions> {
+    kind: Kind,
+    service: &'service Service,
+    application: &'app Application,
+    user: Option<&'user User>,
+    usage: Option<&'usage Usage>,
+    extensions: Option<&'extensions Extensions>,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct Builder<'service, 'app, 'user, 'usage, 'extensions> {
+    service: &'service Service,
+    kind: Option<Kind>,
```

Comment:
Why is `kind` optional? It's required in the struct above. I see that below there are some defaults set. What's the default for `kind`?

Reply:
it's optional because you might have a builder initialized with just the service + credentials but not yet with the kind of call. It's useful to have only as little information as required in the builder and kind is something you can add later on. For example, you can then add some app and usage, clone the builder, and set the kind of the first as Authorize and the second as Report.

Reply:
The default for `kind` is `Option::default()`.

Reply:
:+1: